Refactor `repr` and `str` of `QuadraticProgram` and `OptimizationResult` (part 3/3 refactor text representation of objects)

t-imamichi opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

- Adds `qiskit_optimization.algorithms.OptimizationResult.prettyprint` to display a multi-line text.
- Adds `wrap` option to `prettyprint` function and method to set the text width.
- Updates `str` and `repr` of `QuadraticProgram` and `OptimizationResult` to be a one-line text.
- Reword "optimal" in `OptimizationResult` and `OptimizationResultStatus` because some algorithms do not guarantee the optimality.
  - "optimal solution" (`OptimizationResultStatus.SUCCESS`) -> "feasible solution"
  - "optimal function value" (`OptimizationResult.fval`) -> "objective function value"
  - "optimal value" (`OptimizationResult.x`) -> "variable values"


TODO
- [x] Refactors `repr` and `str` of `QuadraticProgram` and `OptimizationResult` to be one line
- [x] Adds `OptimizationResult.prettyprint` to generate a multi-line string
- [x] Updates docstrings of `OptimizationResults` and `OptimizationResultStatus`
- [x] Updates tutorials to replace `export_as_lp_string` with `prettyprint`
- [x] Updates reno 

Closes #184
Closes #77

Follow-up of #366 #369

### Details and comments

`QuadraticProgram`
```python
from qiskit_optimization.problems import QuadraticProgram

qp = QuadraticProgram('problem 1')
qp.integer_var(-1, 2, 'x')
qp.integer_var(-1, 2, 'y')
qp.continuous_var(-1, name='z')
qp.minimize(constant=3, linear={'x': 2, 'y': 3}, quadratic={('z', 'x'): -1})
qp.linear_constraint({'x': 1, 'y': -2}, '>=', 2, name='lin_GE')
qp.linear_constraint({'x': 2, 'y': -1}, '==', 1, name='lin_EQ')
qp.quadratic_constraint({'x': 2, 'y': -1}, {('z', 'z'): -1}, '<=', 1)

print(repr(qp))
print()
print(str(qp))
print()
print(qp.prettyprint())
```

```
<QuadraticProgram: minimize -x*z + 2*x + 3*y + 3, 3 variables, 3 constraints, 'problem 1'>

minimize -x*z + 2*x + 3*y + 3 (3 variables, 3 constraints, 'problem 1')

Problem name: problem 1

Minimize
  -x*z + 2*x + 3*y + 3

Subject to
  Linear constraints (2)
    x - 2*y >= 2  'lin_GE'
    2*x - y == 1  'lin_EQ'

  Quadratic constraints (1)
    -z^2 + 2*x - y <= 1  'q0'

  Integer variables (2)
    -1 <= x <= 2
    -1 <= y <= 2

  Continuous variables (1)
    -1 <= z
```
-----------------------

`OptimizationResult`

```python
from qiskit_optimization.problems import QuadraticProgram
from qiskit_optimization.algorithms import OptimizationResult, OptimizationResultStatus

qp = QuadraticProgram()
x = qp.binary_var_list(3)
result = OptimizationResult([1.0,2.0,3.0], 10.0, x, OptimizationResultStatus.SUCCESS)

print(str(result))
print()
print(repr(result))
print()
print(str(result.prettyprint()))
```

```
fval=10.0, x0=1.0, x1=2.0, x2=3.0, status=SUCCESS

<OptimizationResult: fval=10.0, x0=1.0, x1=2.0, x2=3.0, status=SUCCESS>

objective function value: 10.0
variable values: x0=1.0, x1=2.0, x2=3.0
status: SUCCESS
```

woodsp-ibm commented on docs/tutorials/01_quadratic_program.ipynb:
```diff
@@ -63,7 +63,7 @@
    "metadata": {},
    "source": [
-    "You start with an empty model. How to add variables and constraints to a model is explained in the section \"Directly constructing a `QuadraticProgram`\"."
+    "You start with an empty model. How to add variables and constraints to a model is explained in the section [Directly constructing a `QuadraticProgram`](#Directly-constructing-a-QuadraticProgram)."
```
I was viewing this file and the link seemed not to work as expected. It certainly fails to come out in the html documentation where I'd argue its more important there than in the source 
![image](https://user-images.githubusercontent.com/40241007/168169278-c1c9887c-9cbd-4103-9238-096e715850d6.png)

t-imamichi replied:
Thanks. Avoid using the code format in the link (`QuadraticProgram` -> QuadraticProgram) resolves the issue.
![image](https://user-images.githubusercontent.com/31178928/168976745-16c749b3-d6db-44ff-99c6-ceb48a4e0cd7.png)

woodsp-ibm commented on qiskit_optimization/algorithms/optimization_algorithm.py:
```diff
@@ -123,7 +123,7 @@ def __init__(
             if len(x) != len(variables):
                 raise QiskitOptimizationError(
-                    f"Inconsistent size of optimal value and variables. x: size {len(x)} {x}, "
+                    f"Inconsistent size of variables and values. x: size {len(x)} {x}, "
```
In this message it looks like you change `optimal value` to `values` (`variable values` more fully above when describing x) and switched the order of that text.r with `variables`, though in the message itself it has values and variables in that order when it prints out the actual quantities that mismatch

t-imamichi replied:
Thanks. I updated the message as follows. I hope it is comprehensive.
```python
f"Inconsistent size of variable values (x) and variables. x: size {len(x)} {x}, "
f"variables: size {len(variables)} {[v.name for v in variables]}"
```

woodsp-ibm commented on docs/tutorials/02_converters_for_quadratic_programs.ipynb:
```diff
@@ -266,51 +242,32 @@
+      "Subject to\n",
+      "  Linear constraints (2)\n",
+      "    x + xyz_leq@int_slack@0 + 2*xyz_leq@int_slack@1 + 2*xyz_leq@int_slack@2 + y + z@0 + 2*z@1 + 4*z@2 == 5  'xyz_leq'\n",
+      "    x - xyz_geq@int_slack@0 - 2*xyz_geq@int_slack@1 - 3*xyz_geq@int_slack@2 + y + z@0 + 2*z@1 + 4*z@2 == 3  'xyz_geq'\n",
```
These end up wrapping in the html docs so the end result is not as nice. Not sure there is a great deal one can do about this. I imagine that we could have the prettyprint wrap things nicely at a given limit, so it determines/manages the word wrapping, but maybe that's too much work for now.
![image](https://user-images.githubusercontent.com/40241007/168171473-8497d726-552e-470c-8185-507522099f31.png)

t-imamichi replied:
Thanks. Wrapping lines might be a good option to improve the appearance. I will try it.

t-imamichi replied:
I'm trying `wrap` option. Here are some results with different wrap values. What do you think of these appearance? If OK, I will write docstrings  and unit tests.

wrap = 100
```
Problem name: problem 1

Minimize
   43*lin_GE@int_slack@0^2 + 86*lin_GE@int_slack@0*lin_GE@int_slack@1 + 43*lin_GE@int_slack@1^2
   - 86*x@0*lin_GE@int_slack@0 - 86*x@0*lin_GE@int_slack@1 + 215*x@0^2 + 860*x@0*x@1 - 344*x@0*y@0
   - 688*x@0*y@1 - x@0*z@0 - 2*x@0*z@1 - 3*x@0*z@2 - 172*x@1*lin_GE@int_slack@0 - 172*x@1*lin_GE@int_slack@1
   + 860*x@1^2 - 688*x@1*y@0 - 1376*x@1*y@1 - 2*x@1*z@0 - 4*x@1*z@1 - 6*x@1*z@2 + 172*y@0*lin_GE@int_slack@0
   + 172*y@0*lin_GE@int_slack@1 + 215*y@0^2 + 860*y@0*y@1 + 344*y@1*lin_GE@int_slack@0
   + 344*y@1*lin_GE@int_slack@1 + 860*y@1^2 + 86*lin_GE@int_slack@0 + 86*lin_GE@int_slack@1 - 427*x@0
   - 854*x@1 + 347*y@0 + 694*y@1 + z@0 + 2*z@1 + 3*z@2 + 212 

Subject to
  No constraints

  Binary variables (12)
     x@0 x@1 y@0 y@1 z@0 z@1 z@2 xxx3 xxx4 xxx5 lin_GE@int_slack@0 lin_GE@int_slack@1
```

wrap = 50
```
Problem name: problem 1

Minimize
   43*lin_GE@int_slack@0^2
   + 86*lin_GE@int_slack@0*lin_GE@int_slack@1
   + 43*lin_GE@int_slack@1^2
   - 86*x@0*lin_GE@int_slack@0
   - 86*x@0*lin_GE@int_slack@1 + 215*x@0^2
   + 860*x@0*x@1 - 344*x@0*y@0 - 688*x@0*y@1 - x@0*z@0
   - 2*x@0*z@1 - 3*x@0*z@2 - 172*x@1*lin_GE@int_slack@0
   - 172*x@1*lin_GE@int_slack@1 + 860*x@1^2
   - 688*x@1*y@0 - 1376*x@1*y@1 - 2*x@1*z@0 - 4*x@1*z@1
   - 6*x@1*z@2 + 172*y@0*lin_GE@int_slack@0
   + 172*y@0*lin_GE@int_slack@1 + 215*y@0^2
   + 860*y@0*y@1 + 344*y@1*lin_GE@int_slack@0
   + 344*y@1*lin_GE@int_slack@1 + 860*y@1^2
   + 86*lin_GE@int_slack@0 + 86*lin_GE@int_slack@1
   - 427*x@0 - 854*x@1 + 347*y@0 + 694*y@1 + z@0 + 2*z@1
   + 3*z@2 + 212 

Subject to
  No constraints

  Binary variables (12)
     x@0 x@1 y@0 y@1 z@0 z@1 z@2 xxx3 xxx4 xxx5
     lin_GE@int_slack@0 lin_GE@int_slack@1
```

woodsp-ibm replied:
Looks good to me. Given you show different values above I assume you will wrap at some default value and let the user change that if they want for some reason.

t-imamichi replied:
Thanks, I will tune the default value looking at the jupyter notebook output and let users change that if they wish.

t-imamichi commented:
I added a text wrap feature to `prettyprint` and set text width 80 as the default value.
I implemented a simple approach to wrap texts at `+` or `-` only. It does not wrap text at `*` and (in)equalities.
So, the `wrap` value is not a hard constraint (i.e., some text may exceed).
But, I think it's enough comprehensive. See https://github.com/Qiskit/qiskit-optimization/pull/370#discussion_r873707634 for examples.

woodsp-ibm reviewed (approved):
LGTM thx!

t-imamichi commented:
Thank you for your reviews of these PR series! I finally completed them.
